reader-shp/src/main/java/com/graphhopper/reader/shp/OSMShapeFileReader.java

```diff
@@ -0,0 +1,329 @@
+						tmpSet.clear();
+						for (int i = 0; i < points.length; i++) {
+							Coordinate c = points[i];
+
+							// don't add the same coord twice for the same edge - happens with bad geometry, i.e.
+							// duplicate coords or a road which forms a circle (e.g. roundabout)
+							if (tmpSet.contains(c)) {
+								continue;
+							}
+							tmpSet.add(c);
+
+							// skip if its already a node
+							int state = coordState.get(c);
+							if (state >= FIRST_NODE_ID) {
+								continue;
+							}
+
+							if (i == 0 || i == points.length - 1 || state == COORD_STATE_PILLAR) {
+								// turn into a node if its the first or last point, or already appeared in another edge
+								int nodeId = nextNodeId++;
+								coordState.put(c, nodeId);
+								saveTowerPosition(nodeId, c);
+							} else if (state == COORD_STATE_UNKNOWN) {
+								// mark it as a pillar (which may get upgraded to an edge later)
+								coordState.put(c, COORD_STATE_PILLAR);
+							}
+						}
+					}
+
+				}
+			} finally {
+				if (roads != null) {
+					roads.close();
+				}
+				if (dataStore != null) {
+					dataStore.dispose();
+				}
+			}
+
+			LOGGER.info("Number of junction points : " + (nextNodeId - FIRST_NODE_ID));
+
+		}
+	}
+
+	@Override
+	void processRoads() throws IOException {
+
+		DataStore dataStore = null;
+		FeatureIterator<SimpleFeature> roads = null;
+
+		try {
+			dataStore = openShapefileDataStore(roadsFile);
+			roads = getFeatureIerator(dataStore);
+
+			while (roads.hasNext()) {
+				SimpleFeature road = roads.next();
+
+				for (Coordinate[] points : getCoords(road.getDefaultGeometry())) {
+
+					// Parse all points in the geometry, splitting into individual graphhopper edges
+					// whenever we find a node in the list of points
+					Coordinate startTowerPnt = null;
+					List<Coordinate> pillars = new ArrayList<Coordinate>();
+					for (Coordinate point : points) {
+						if (startTowerPnt == null) {
+							startTowerPnt = point;
+						} else {
+							int state = coordState.get(point);
+							if (state >= FIRST_NODE_ID) {
+								int fromTowerNodeId = coordState.get(startTowerPnt);
+								int toTowerNodeId = state;
+
+								// get distance and estimated centres
+								double distance = getWayLength(startTowerPnt, pillars, point);
+								GHPoint estmCentre = new GHPoint(0.5 * (lat(startTowerPnt) + lat(point)),
+										0.5 * (lng(startTowerPnt) + lng(point)));
+								PointList pillarNodes = new PointList(pillars.size(), false);
+
+								for (Coordinate pillar : pillars) {
+									pillarNodes.add(lat(pillar), lng(pillar));
+								}
+
+								addEdge(fromTowerNodeId, toTowerNodeId, road, distance, estmCentre, pillarNodes);
+								startTowerPnt = point;
+								pillars.clear();
+							} else {
+								pillars.add(point);
+							}
+						}
+					}
+				}
+
+			}
+		} catch (Exception e) {
+			LOGGER.error("Exception while processing for roads: ", e);
```

Comment:
Here we have to throw an exception, e.g. `throw new RuntimeException("some info", e)`